File: triolingopush.cpp
#include <bits/stdc++.h>

using namespace std;

typedef long long ll;

ll MOD = 1000000007;
const int MAX_N = 3;

struct Matrix {ll mat[MAX_N][MAX_N] = {{1,1,1},{1,0,0},{0,0,1}};};


Matrix matMul (Matrix a, Matrix b) {
    Matrix ans;
    for (int i = 0; i < MAX_N; i++) {
        for (int j = 0; j < MAX_N; j++) ans.mat[i][j] = 0;
    }
    for (int i = 0; i < MAX_N; i++) {
        for (int j = 0; j < MAX_N; j++) {
            if (a.mat[i][j] == 0) continue; //optimisation
            for (int k = 0; k < MAX_N; k++) {
                ans.mat[i][k] += (a.mat[i][j] % MOD) * (b.mat[j][k] % MOD);
                ans.mat[i][k] = ans.mat[i][k] % MOD;
            }
        }
    }
    return ans;
}

Matrix matPow(Matrix base, ll p) {
    Matrix ans;
    for (int i = 0; i < MAX_N; i++) {
        for (int j = 0; j < MAX_N; j++) ans.mat[i][j] = (i == j); //prepare identity matrix
    }
    while (p > 0) {
        if (p & 1) ans = matMul(ans,base); // check if p is odd
        base = matMul(base,base);
        p >>= 1;
    }
    return ans;
}

// use matrix exponentiation to find the pth number,denoted as F_{p} in O(log p)
// recursive relation is [1 1 1; 1 0 0; 0 0 1] * [F_{p} F_{p - 1} 1] = [F_{p + 1} F{p} 1]
int main() {
    ll N; scanf("%lld",&N);
    Matrix base;
    base = matPow(base, N - 1); // # need to multiply by N - 1 times
    //printf("res[1][0] = %lld, res[1][1] = %lld, res[1][2] = %lld\n", base.mat[1][0], base.mat[1][1], base.mat[1][2]);
    printf("%lld\n", ((base.mat[1][0] * 2) % MOD + (base.mat[1][1] + base.mat[1][2]) % MOD) % MOD);
    return 0;
}

File: triolingopush.py
MAX_N,MOD = 3,pow(10,9) + 7
BASE = [[1,1,1],[1,0,0],[0,0,1]]

def mod(a,m):
    return ((a % m) + m) % m

def mat_mul(A,B):
    #assume matrix multiplication is possible
    ans = [[0 for i in range(MAX_N)] for j in range(MAX_N)]
    for i in range(MAX_N):
        for j in range(MAX_N):
            if A[i][j] == 0: continue
            for k in range(MAX_N): 
                ans[i][k] += mod(A[i][j], MOD) * mod(B[j][k], MOD);
                ans[i][k] = mod(ans[i][k], MOD);
    return ans

def eye(m):
    identity = [[0 for i in range(m)] for j in range(m)]
    for i, row in enumerate(identity):
        row[i] = 1
    return identity

def mat_pow(mat, power):
    result = eye(MAX_N)
    if power == 0:
        return result
    while power > 0:
        if power & 1 == 1:
            result = mat_mul(result, mat)
        mat = mat_mul(mat, mat)
        power >>= 1
    return result

N = int(input())
res = mat_pow(BASE, N - 1)
print(mod((res[1][0] * 2 + res[1][1] + res[1][2]),MOD))
